### PR TITLE
MOB-21374 | Get proposition call will be dispatched immediately if all scopes are cached.

### PR DESCRIPTION
**Pull request by ishwetansh**
Use Case 1: If getPropositions() is called for a decision scope (e.g., scope1) and there's an updatePropositions() in progress for a different scope (e.g., scope2, scope3), return the value for scope1 immediately without waiting.

Use Case 2: If getPropositions() is called for a decision scope that is currently being updated (e.g., scope1), wait for the updatePropositions() to complete before returning the response.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by spoorthipujariadobe — COMMENTED** [review]
Logic looks good. As discussed offline, the test is failing because of `handleGetPropositions` being called from SerialDispatcher which runs on a different thread and does not get the mocked Based64.class instance.
As a fix let's move these tests into OptimizeFunctionalTests
1. testHandleOptimizeRequestContent_GetPropositionsEvent_whenUpdateIsComplete
2. testGetPropositions_dispatchPropositionFromCacheBeforeNextUpdate
3. testGetPropositions_ScopesFromEventIsUpdateInProgress

And add this test to `OptimizeExtensionTests` to unit test `handleGetPropositions`

1. `testHandleOptimizeRequestContent_GetPropositionsEvent_propositionInCache`: Make sure update is not called before get. Directly add the proposition in `cachedPropositions` and call `handleOptimizeRequestContent` with a get event for the cached proposition

File: code/optimize/src/androidTest/java/com/adobe/marketing/mobile/optimize/OptimizeFunctionalTests.java [hunk 1]
```diff
@@ -2194,6 +2194,432 @@ public void testPropositionGenerateReferenceXdm() throws IOException {
                 "de03ac85-802a-4331-a905-a57053164d35", decisioning.get("propositionID"));
     }
 
+    // 19
+    @Test
+    public void testGetPropositions_ScopesFromEventIsUpdateInProgress()
```
Comment (spoorthipujariadobe) [inline code/optimize/src/androidTest/java/com/adobe/marketing/mobile/optimize/OptimizeFunctionalTests.java:2199]: nit: Can we name is to describe what happens in the test? `testGetPropositions_multipleUpdatePropositonsCallsBeforeGetPropositions`

**Review by spoorthipujariadobe — APPROVED** [review]
LGTM! Small nit about naming of test

File: code/optimize/src/androidTest/java/com/adobe/marketing/mobile/optimize/OptimizeFunctionalTests.java (hunk 1 above)
Comment (ishwetansh) [reply code/optimize/src/androidTest/java/com/adobe/marketing/mobile/optimize/OptimizeFunctionalTests.java]: Sure, makes more sense 👍

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java [hunk 2]
```diff
@@ -232,10 +233,78 @@ void handleOptimizeRequestContent(@NonNull final Event event) {
...
+                                "handleOptimizeRequestContent - Scopes are not fully cached or are"
+                                        + " in progress, adding event to dispatcher.");
+                        eventsDispatcher.offer(event);
+                    }
+                    break;
+                } catch (final Exception e) {
+                    Log.warning(
+                            OptimizeConstants.LOG_TAG,
+                            SELF_TAG,
+                            "handleUpdatePropositions - Failed to process update propositions"
```
Comment (spoorthipujariadobe) [inline code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:304]: nit: this needs to be `"handleOptimizeRequestContent - Failed to process get propositions"`

File: code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java [hunk 3]
```diff
@@ -232,10 +233,78 @@ void handleOptimizeRequestContent(@NonNull final Event event) {
...
+                                    eventData,
+                                    OptimizeConstants.EventDataKeys.DECISION_SCOPES);
+                    List<DecisionScope> eventDecisionScopes =
+                            retrieveValidDecisionScopes(decisionScopesData);
+
+                    if (OptimizeUtils.isNullOrEmpty(eventDecisionScopes)) {
+                        Log.debug(
+                                OptimizeConstants.LOG_TAG,
+                                SELF_TAG,
+                                "handleGetPropositions - Cannot process the get propositions"
```
Comment (spoorthipujariadobe) [inline code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/OptimizeExtension.java:250]: nit: `handleGetPropositions` -> `handleOptimizeRequestContent`

**Review by spoorthipujariadobe — APPROVED** [review]

**Pull request merged**